File: lib/test/watch_test.lua
-- Copyright 2023 The Yock Authors. All rights reserved.
-- Use of this source code is governed by a MIT-style
-- license that can be found in the LICENSE file.

job("cpu", function(cenv)
    table.dump(cpu.percent(3 * time.Second, true))
    table.dump(cpu.times(true))
    table.dump(cpu.times(false))
    table.dump(cpu)
end)

job("disk", function(cenv)
    table.dump(disk.info())
    table.dump(disk.partitions(true))
    table.dump(disk.usage("D:"))
end)

job("mem", function(cenv)
    table.dump(mem.swap())
    table.dump(mem.info())
end)

job("host", function(cenv)
    print(host.boot_time())
    print(host.info())
end)

job("net", function(cenv)
    table.dump(net.io(true))
    table.dump(net.interfaces())
end)

jobs("all", "cpu", "disk", "mem", "host", "net")

File: watch/metrics/watch_test.go
// Copyright 2023 The Yock Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package metrics

import (
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"testing"
	"time"

	"github.com/ansurfen/yock/util"
	"github.com/prometheus/client_golang/prometheus"
)

func TestMetricsWatch(t *testing.T) {
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)

	DefaultMetricsWatch = New()

	go DefaultMetricsWatch.Snapshot()

	go func() {
		for {
			time.Sleep(5 * time.Second)
			name := util.RandString(32)
			c := DefaultMetricsWatch.NewCounter(prometheus.CounterOpts{
				Name: name,
			})
			go func() {
				for {
					c.Inc()
					time.Sleep(2 * time.Second)
				}
			}()
		}
	}()

	http.Handle("/metrics", DefaultMetricsWatch.Document())

	go http.ListenAndServe(":2112", nil)
	<-c
}
